src/ctap/apdu.rs

```diff
@@ -0,0 +1,303 @@
+            ApduStatusCode::SW_FILE_NOT_FOUND => b"\x6a\x82",
+            ApduStatusCode::SW_INCORRECT_P1P2 => b"\x6a\x86",
+            ApduStatusCode::SW_INS_INVALID => b"\x6d\x00",
+            ApduStatusCode::SW_CLA_INVALID => b"\x6e\x00",
+            ApduStatusCode::SW_INTERNAL_EXCEPTION => b"\x6f\x00",
+        }
+    }
+}
+
+#[allow(non_camel_case_types, dead_code)]
+pub enum ApduIns {
+    SELECT = 0xA4,
+    READ_BINARY = 0xB0,
+    GET_RESPONSE = 0xC0,
+}
+
+#[cfg_attr(test, derive(Clone, Debug))]
+#[allow(dead_code)]
+#[derive(Default, PartialEq)]
+pub struct ApduHeader {
+    cla: u8,
+    ins: u8,
+    p1: u8,
+    p2: u8,
+}
+
+impl From<&[u8]> for ApduHeader {
+    fn from(header: &[u8]) -> Self {
+        ApduHeader {
+            cla: header[0],
+            ins: header[1],
+            p1: header[2],
+            p2: header[3],
+        }
+    }
+}
+
+#[cfg_attr(test, derive(Clone, Debug))]
+#[allow(dead_code)]
+#[derive(Default, PartialEq)]
+pub struct APDU {
+    header: ApduHeader,
+    lc: u16,
+    data: Vec<u8>,
+    le: u32,
+    case_type: u8,
+}
+
+const APDU_HEADER_LEN: u8 = 4;
+
+impl TryFrom<&[u8]> for APDU {
+    type Error = ApduStatusCode;
+
+    fn try_from(frame: &[u8]) -> Result<Self, ApduStatusCode> {
+        if frame.len() < APDU_HEADER_LEN as usize {
+            return Err(ApduStatusCode::SW_WRONG_DATA);
+        }
+        //        +-----+-----+----+----+
+        // header | CLA | INS | P1 | P2 |
+        //        +-----+-----+----+----+
+        let (header, payload) = frame.split_at(APDU_HEADER_LEN as usize);
+
+        let mut apdu = APDU {
+            header: header.into(),
+            lc: 0x00,
+            data: Vec::new(),
+            le: 0x00,
+            case_type: 0x00,
+        };
+
+        // case 1
+        if payload.is_empty() {
+            apdu.case_type = 0x01;
+        } else {
+            let byte_0 = payload[0];
+            // case 2S (Le)
+            if payload.len() == 1 {
+                apdu.case_type = 0x02;
+                apdu.le = if byte_0 == 0x00 {
+                    // Ne = 256
+                    0x100
+                } else {
+                    byte_0.into()
+                }
+            }
+            // case 3S (Lc + data)
+            if payload.len() == (1 + byte_0) as usize && byte_0 != 0 {
+                apdu.case_type = 0x03;
+                apdu.lc = byte_0.into();
+                apdu.data = payload[1..].to_vec();
+            }
+            // case 4S (Lc + data + Le)
+            if payload.len() == (1 + byte_0 + 1) as usize && byte_0 != 0 {
+                apdu.case_type = 0x04;
+                apdu.lc = byte_0.into();
+                apdu.data = payload[1..(payload.len() - 1)].to_vec();
+                apdu.le = (*payload.last().unwrap()).into();
+                if apdu.le == 0x00 {
+                    apdu.le = 0x100;
+                }
+            }
+        }
+        // TODO: Add extended length cases
```

Comment:
Which NFC issue are you referring to?

Reply:
Hmm, the open NFC issue is a bit to broad for this piece of code. Also, is extended length compatible with our drivers?

Reply:
The existing [parser](https://github.com/google/OpenSK/blob/f204005ecaa3410fb03e0d04cd9c98b6f0950de7/src/ctap/ctap1.rs#L114) works only with extended length APDUs. So I think if this module would be used in the future either with NFC or USB, then it will support both APDU types 🤔

Reply:
I think leaving it open is fine.